### ISSUE TITLE
StreamClosedError when receiving RST_STREAM on a stream closed by remote

### ISSUE THREAD

**Issue by standy66** [python-hyper/hyper-h2#1183]
I am trying to make h2 state machine behave itself correctly when communicating with gRPC Core peers. Right now h2 raises `StreamClosedError` when receiving *RST_STREAM* on a closed stream that was closed by remote, e.g.

```
# Local (h2) -> Remote (gRPC Core)
-> HEADERS frame
-> DATA frame with END_STREAM flag set  # stream is now half-closed (local)
<- HEADERS frame
<- DATA frame
<- HEADERS frame with END_STREAM flag set  # stream is now closed
<- RST_STREAM
```

Unfortunately, this is what gRPC Core does for unary RPC calls, because it cannot accept more data than a single message, delivered in one or possibly multiple *DATA* frames. I am not sure if it's prohibited the the RFC or not, but it is what it is. The workaround I am using for now is to [monkey patch h2's stream transition dictionary](https://github.com/standy66/purerpc/blob/anyio/src/purerpc/grpclib/connection.py#L21) so it ignores any future *RST_STREAM* and *WINDOW_UPDATE* events on a remotely closed stream.

**Comment by pgjones** [comment]
I think the RFC does allow a RST_STREAM frame after an END_STREAM, based on this [lifecycle](https://tools.ietf.org/html/rfc7540#section-5.1) diagram. 

(I'll put this here to see if someone more knowledgeable corrects me, and if they don't I'll try to fix this the next time I work on this library.)

**Comment by standy66** [comment]
Here is an excerpt from RFC about closed state:

> Endpoints MUST ignore
>       WINDOW_UPDATE or RST_STREAM frames received in this state, though
>       endpoints MAY choose to treat frames that arrive a significant
>       time after sending END_STREAM as a connection error
>       (Section 5.4.1) of type PROTOCOL_ERROR

Right now I think h2 ignores WINDOW_UPDATE or RST_STREAM frames on a closed stream only if they originated from a party other than who sent the END_STREAM flag. What should happen is that it must ignore those frames regardless.

**Comment by pgjones** [comment]
Fixed with a95556b95dd21c337312658a5570c48f5e13abee
